File: src/main/java/de/pflugradts/passbird/adapter/passwordstore/PasswordStoreWriter.java
package de.pflugradts.passbird.adapter.passwordstore;

import com.google.inject.Inject;
import de.pflugradts.passbird.application.configuration.ReadableConfiguration;
import de.pflugradts.passbird.application.failurehandling.FailureCollector;
import de.pflugradts.passbird.application.util.ByteArrayUtils;
import de.pflugradts.passbird.application.util.SystemOperation;
import de.pflugradts.passbird.domain.model.namespace.Namespace;
import de.pflugradts.passbird.domain.model.namespace.NamespaceSlot;
import de.pflugradts.passbird.domain.model.password.PasswordEntry;
import de.pflugradts.passbird.domain.model.transfer.Bytes;
import de.pflugradts.passbird.domain.service.NamespaceService;
import de.pflugradts.passbird.domain.service.password.encryption.CryptoProvider;
import io.vavr.control.Try;
import lombok.AllArgsConstructor;
import lombok.NoArgsConstructor;

import java.nio.file.Path;
import java.util.Arrays;
import java.util.Optional;
import java.util.function.Supplier;
import java.util.stream.Collectors;
import java.util.stream.Stream;
import java.util.stream.StreamSupport;

import static de.pflugradts.passbird.application.configuration.ReadableConfiguration.DATABASE_FILENAME;
import static de.pflugradts.passbird.domain.model.namespace.NamespaceSlot.CAPACITY;
import static de.pflugradts.passbird.domain.model.namespace.NamespaceSlot.FIRST_NAMESPACE;
import static de.pflugradts.passbird.domain.model.namespace.NamespaceSlot.LAST_NAMESPACE;

@NoArgsConstructor
@AllArgsConstructor
class PasswordStoreWriter {

    private static final Integer EOF = 0;
    private static final Integer SECTOR = -1;

    @Inject
    private SystemOperation systemOperation;
    @Inject
    private FailureCollector failureCollector;
    @Inject
    private ReadableConfiguration configuration;
    @Inject
    private PasswordEntryTransformer passwordEntryTransformer;
    @Inject
    private NamespaceTransformer namespaceTransformer;
    @Inject
    private NamespaceService namespaceService;
    @Inject
    private CryptoProvider cryptoProvider;
    @Inject
    private PasswordStoreCommons commons;

    public void sync(final Supplier<Stream<PasswordEntry>> passwordEntriesSupplier) {
        final var contentSize = calcRequiredContentSize(passwordEntriesSupplier);
        final var bytes = new byte[calcActualTotalSize(contentSize)];
        var offset = ByteArrayUtils.copyBytes(commons.signature(), bytes, 0, commons.signatureSize());
        offset = persistNamespaces(bytes, offset);
        for (final PasswordEntry passwordEntry : passwordEntriesSupplier.get().collect(Collectors.toList())) {
            final var entryBytes = passwordEntryTransformer.transform(passwordEntry);
            offset += ByteArrayUtils.copyBytes(entryBytes, bytes, offset, entryBytes.length);
        }
        offset += ByteArrayUtils.copyBytes(EOF, bytes, offset);
        final byte[] checksumBytes = {contentSize > 0
                ? checksum(Arrays.copyOfRange(bytes, commons.signatureSize(), contentSize))
                : 0x0};
        ByteArrayUtils.copyBytes(checksumBytes, bytes, offset, commons.checksumBytes());
        writeToDisk(Bytes.bytesOf(bytes))
            .onFailure(throwable ->
                failureCollector.collectWritePasswordDatabaseFailure(getFilePath(), throwable));
    }

    private int persistNamespaces(final byte[] bytes, final int offset) {
        var incrementedOffset = offset;
        ByteArrayUtils.copyBytes(SECTOR, bytes, incrementedOffset);
        incrementedOffset += commons.intBytes();
        for (int index = FIRST_NAMESPACE; index <= LAST_NAMESPACE; index++) {
            final var namespaceBytes = namespaceTransformer.transform(NamespaceSlot.at(index));
            incrementedOffset += ByteArrayUtils.copyBytes(
                namespaceBytes, bytes, incrementedOffset, namespaceBytes.length);
        }
        return incrementedOffset;
    }

    private Try<Path> writeToDisk(final Bytes bytes) {
        return systemOperation.writeBytesToFile(
            getFilePath(),
            cryptoProvider.encrypt(bytes).getOrElse(Bytes.emptyBytes()));
    }

    private byte checksum(final byte[] bytes) {
        return (byte) StreamSupport.stream(Bytes.bytesOf(bytes).spliterator(), false)
            .mapToInt(b -> (int) (byte) b)
            .reduce(0, Integer::sum);
    }

    private int calcRequiredContentSize(final Supplier<Stream<PasswordEntry>> passwordEntries) {
        final var dataSize = passwordEntries.get()
            .map(passwordEntry ->
                commons.intBytes() + passwordEntry.viewKey().getSize() + passwordEntry.viewPassword().getSize())
            .reduce(0, Integer::sum);
        final var namespaceSize = commons.intBytes() + CAPACITY * commons.intBytes() + namespaceService.all()
            .filter(Optional::isPresent)
            .map(Optional::get)
            .map(Namespace::getBytes)
            .map(Bytes::getSize)
            .reduce(0, Integer::sum);
        final var metaSize = (int) passwordEntries.get().count() * 2 * commons.intBytes();
        return dataSize + namespaceSize + metaSize;
    }

    private Path getFilePath() {
        return systemOperation.resolvePath(
            configuration.getAdapter().getPasswordStore().getLocation(),
            DATABASE_FILENAME
        ).getOrNull();
    }

    private int calcActualTotalSize(final int contentSize) {
        return commons.signatureSize() + contentSize + commons.eofBytes() + commons.checksumBytes();
    }

}

File: src/main/kotlin/de/pflugradts/passbird/adapter/passwordstore/PasswordStoreWriter.kt
package de.pflugradts.passbird.adapter.passwordstore

import com.google.inject.Inject
import de.pflugradts.kotlinextensions.tryCatching
import de.pflugradts.passbird.application.configuration.ReadableConfiguration
import de.pflugradts.passbird.application.configuration.ReadableConfiguration.Companion.DATABASE_FILENAME
import de.pflugradts.passbird.application.failure.WritePasswordDatabaseFailure
import de.pflugradts.passbird.application.failure.reportFailure
import de.pflugradts.passbird.application.toDirectory
import de.pflugradts.passbird.application.toFileName
import de.pflugradts.passbird.application.util.SystemOperation
import de.pflugradts.passbird.application.util.copyBytes
import de.pflugradts.passbird.application.util.copyInt
import de.pflugradts.passbird.domain.model.egg.Egg
import de.pflugradts.passbird.domain.model.nest.NestSlot
import de.pflugradts.passbird.domain.model.shell.Shell
import de.pflugradts.passbird.domain.model.shell.Shell.Companion.emptyShell
import de.pflugradts.passbird.domain.model.shell.Shell.Companion.shellOf
import de.pflugradts.passbird.domain.service.NestService
import de.pflugradts.passbird.domain.service.password.encryption.CryptoProvider
import de.pflugradts.passbird.domain.service.password.storage.EggStreamSupplier
import java.util.Arrays

private const val EOF = 0
private const val SECTOR = -1
private const val EMPTY_NEST = -2

class PasswordStoreWriter @Inject constructor(
    @Inject private val systemOperation: SystemOperation,
    @Inject private val configuration: ReadableConfiguration,
    @Inject private val nestService: NestService,
    @Inject private val cryptoProvider: CryptoProvider,
) {

    fun sync(eggSupplier: EggStreamSupplier) {
        val contentSize = calcRequiredContentSize(eggSupplier)
        val bytes = ByteArray(calcActualTotalSize(contentSize))
        var offset = copyBytes(signature(), bytes, 0, signatureSize())
        offset = persistNest(bytes, offset)
        eggSupplier.get().forEach { egg -> egg.asByteArray().let { offset += copyBytes(it, bytes, offset, it.size) } }
        offset += copyInt(EOF, bytes, offset)
        val checksumBytes = byteArrayOf(if (contentSize > 0) checksum(Arrays.copyOfRange(bytes, signatureSize(), contentSize)) else 0x0)
        copyBytes(checksumBytes, bytes, offset, checksumBytes())
        writeToDisk(shellOf(bytes))
    }

    private fun persistNest(bytes: ByteArray, offset: Int): Int {
        var incrementedOffset = offset
        copyInt(SECTOR, bytes, incrementedOffset)
        incrementedOffset += intBytes()
        for (index in NestSlot.FIRST_SLOT..NestSlot.LAST_SLOT) {
            NestSlot.at(index).asByteArray().let { incrementedOffset += copyBytes(it, bytes, incrementedOffset, it.size) }
        }
        return incrementedOffset
    }

    private fun writeToDisk(shell: Shell) =
        tryCatching { systemOperation.writeBytesToFile(filePath, cryptoProvider.encrypt(shell)) }
            .onFailure { reportFailure(WritePasswordDatabaseFailure(filePath, it)) }

    private fun calcRequiredContentSize(eggs: EggStreamSupplier): Int {
        val dataSize = eggs.get()
            .map { egg: Egg -> intBytes() + egg.viewEggId().size + egg.viewPassword().size }
            .reduce(0) { a: Int, b: Int -> Integer.sum(a, b) }
        val nestSize = intBytes() + NestSlot.CAPACITY * intBytes() + nestService.all()
            .filter { it.isPresent }
            .map { it.get().shell.size }
            .reduce(0) { a: Int, b: Int -> Integer.sum(a, b) }
        val metaSize = eggs.get().count().toInt() * 2 * intBytes()
        return dataSize + nestSize + metaSize
    }

    private val filePath get() =
        systemOperation.resolvePath(configuration.adapter.passwordStore.location.toDirectory(), DATABASE_FILENAME.toFileName())
    private fun calcActualTotalSize(contentSize: Int) = signatureSize() + contentSize + eofBytes() + checksumBytes()

    private fun NestSlot.asByteArray(): ByteArray {
        val nestShell = nestService.atNestSlot(this).map { it.shell }.orElse(emptyShell())
        val nestBytesSize = nestShell.size
        val bytes = ByteArray(Integer.BYTES + nestBytesSize)
        copyInt(if (nestShell.isEmpty) EMPTY_NEST else nestBytesSize, bytes, 0)
        if (!nestShell.isEmpty) { copyBytes(nestShell.toByteArray(), bytes, Integer.BYTES, nestBytesSize) }
        return bytes
    }

    private fun Egg.asByteArray(): ByteArray {
        val eggIdSize = viewEggId().size
        val passwordSize = viewPassword().size
        val metaSize = 2 * Integer.BYTES
        val bytes = ByteArray(Integer.BYTES + eggIdSize + passwordSize + metaSize)
        var offset = copyInt(associatedNest().index(), bytes, 0)
        offset += copyInt(eggIdSize, bytes, offset)
        offset += copyBytes(viewEggId().toByteArray(), bytes, offset, eggIdSize)
        offset += copyInt(passwordSize, bytes, offset)
        copyBytes(viewPassword().toByteArray(), bytes, offset, passwordSize)
        return bytes
    }
}
